File: chessbot/engine/perft.py
"""
perft.py - Performance test (perft) for move generation correctness.

Perft counts the number of leaf nodes at a given depth to verify
move generation is correct.
"""

from typing import Dict
from .board import Board
from .legal import generate_legal_moves


def perft(board: Board, depth: int) -> int:
    """
    Count leaf nodes at a given depth.
    """
    if depth == 0:
        return 1

    moves = generate_legal_moves(board)
    nodes = 0

    for move in moves:
        # Make move on a copy
        board_copy = board.copy()
        board_copy.make_move(move)

        # Recursively count nodes
        nodes += perft(board_copy, depth - 1)

    return nodes


def perft_divide(board: Board, depth: int) -> Dict[str, int]:
    """
    Perft with divide - shows node count for each root move.

    Returns:
        Dictionary mapping move strings to node counts
    """
    if depth == 0:
        return {}

    moves = generate_legal_moves(board)
    results = {}
    total = 0

    for move in moves:
        # Make move on a copy
        board_copy = board.copy()
        board_copy.make_move(move)

        # Count nodes for this move
        count = perft(board_copy, depth - 1)
        results[str(move)] = count
        total += count

    # Add total
    results['TOTAL'] = total

    return results

File: engine/src/main/scala/chesslab/core/Perft.scala
package chesslab.core

object Perft:

  def perft(bb: BitBoard, depth: Int): Long =
    if depth == 0 then 1L
    else
      val moves = BitLegal.legalMoves(bb)
      if depth == 1 then moves.size.toLong
      else moves.foldLeft(0L) { (nodes, move) =>
        nodes + perft(bb.makeMove(move), depth - 1)
      }

  def divide(bb: BitBoard, depth: Int): Map[String, Long] =
    if depth == 0 then Map.empty
    else
      val results = BitLegal.legalMoves(bb).map { move =>
        FenCodec.moveToUci(move) -> perft(bb.makeMove(move), depth - 1)
      }.toMap
      val total = results.values.foldLeft(0L)(_ + _)
      results + ("TOTAL" -> total)
